prusti-viper/src/encoder/array_encoder.rs

```diff
@@ -0,0 +1,237 @@
+    vir,
+    vir_local,
+};
+
+
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub struct EncodedArrayTypes<'tcx> {
+    pub array_pred: String,
+    pub array_ty: vir::Type,
+    pub elem_ty: vir::Type,
+    pub elem_value_ty: vir::Type,
+    pub elem_ty_rs: ty::Ty<'tcx>,
+    pub array_len: usize,
+    pub lookup_pure_name: String,
+}
+
+impl<'tcx> EncodedArrayTypes<'tcx> {
+    pub fn encode_lookup_pure_call(&self, array: vir::Expr, idx: vir::Expr) -> vir::Expr {
+        vir::Expr::func_app(
+            self.lookup_pure_name.clone(),
+            vec![
+                array,
+                idx,
+            ],
+            vec![
+                vir_local!{ self: {self.array_ty.clone()} },
+                vir_local!{ idx: Int },
+            ],
+            self.elem_value_ty.clone(),
+            vir::Position::default(),
+        )
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub struct EncodedSliceTypes<'tcx> {
+    pub slice_pred: String,
+    pub slice_ty: vir::Type,
+    pub elem_ty: vir::Type,
+    pub elem_value_ty: vir::Type,
+    pub elem_ty_rs: ty::Ty<'tcx>,
+    pub lookup_pure_name: String,
+    pub slice_len_name: String,
+}
+
+impl<'tcx> EncodedSliceTypes<'tcx> {
+    pub fn encode_lookup_pure_call(&self, slice: vir::Expr, idx: vir::Expr) -> vir::Expr {
+        vir::Expr::func_app(
+            self.lookup_pure_name.clone(),
+            vec![
+                slice,
+                idx,
+            ],
+            vec![
+                vir_local!{ self: {self.slice_ty.clone()} },
+                vir_local!{ idx: Int },
+            ],
+            self.elem_value_ty.clone(),
+            vir::Position::default(),
+        )
+    }
+
+    pub fn encode_slice_len_call(&self, slice: vir::Expr) -> vir::Expr {
+        vir::Expr::func_app(
+            self.slice_len_name.clone(),
+            vec![
+                slice,
+            ],
+            vec![
+                vir_local!{ self: {self.slice_ty.clone()} },
+            ],
+            vir::Type::Int,
+            vir::Position::default(),
+        )
+    }
+}
+
+
+
+pub struct ArrayTypesEncoder<'tcx> {
+    array_types_cache: HashMap<ty::Ty<'tcx>, EncodedArrayTypes<'tcx>>,
+    slice_types_cache: HashMap<ty::Ty<'tcx>, EncodedSliceTypes<'tcx>>,
+}
+
+impl<'p, 'v: 'p, 'tcx: 'v> ArrayTypesEncoder<'tcx> {
+    pub fn new() -> Self {
+        ArrayTypesEncoder {
+            array_types_cache: HashMap::new(),
+            slice_types_cache: HashMap::new(),
+        }
+    }
+
+    /// Encode types, type predicates and builtin lookup functions required for handling slices of
+    /// type `slice_ty`.
+    ///
+    /// Optionally pass a pre-encoded `vir::Type` for the element snapshot type. This is useful
+    /// when calling `encode_slice_types` from e.g. inside the `SnapshotEncoder` where `Encoder`'s
+    /// `snapshot_encoder` field is already borrowed, so `encoder.encode_snapshot_type` would
+    /// result in a double borrow (i.e. panic).
```

Comment:
I find this `pre_encoded_elem_snap_ty` API a bit messy. How does `SnapshotEncoder` call this?

Reply:
it is messy indeed. one alternative i thought about is to drop that field from the struct, but we need it in `encode_lookup_pure_call`.

SnapshotEncoder is (sometimes/mostly) called through [`Encoder::encode_snapshot_type`](https://github.com/viperproject/prusti-dev/blob/master/prusti-viper/src/encoder/encoder.rs#L881), which does `        self.snapshot_encoder.borrow_mut().encode_type(self, ty)`. The `SnapshotEncoder::encode_type` call goes through pretty much the whole snapshot machinery for the type, building the required functions and axioms and whatnot, and in the end only returning the type of the snapshot. If anything inside that machinery calls `encode_array_types`, then that would arrive at the same `self.snapshot_encoder.borrow_mut()` again, making prusti panic. So instead, `SnapshotEncoder` would just call `self.encode_type` on the array element type first, and use the result of that, without going through `Encoder::encode_snapshot_type` thus avoiding the borrowing problem.

I'm not sure if we can make that `borrow_mut` more fine-grained by moving the `Rc` into `SnapshotEncoder` (so we'd only need a shared reference for the calls from `Encoder`, or if that would just move the problem elsewhere. i hope it would drastically shorten the duration (and split it up by its fields) of the `SnapshotEncoder` are borrowed).

Reply:
hm, now that i think about it.. we could make it a little less messy by dropping the element snapshot type from the `EncodedArrayTypes` struct and requiring the user to pass it to `EncodedArrayTypes::encode_lookup_pure_call`. That way, we can still pass whatever is appropriate/possible in the current situation, but we don't need the admittedly ugly API to thread this additional (not really intuitive) parameter through.